### ISSUE TITLE
nonce, request.csp_nonce and {% script %} all fail to render a nonce

### ISSUE THREAD

**Issue by Flimm** [mozilla/django-csp#167]
I can't get the `nonce` value to be included in the HTML. I've tried adding the context processor, and adding the app to `INSTALLED_APPS`. I've tried:

```
<script nonce="{{ nonce }}">
```

as well as:
```
<script nonce="{{ request.csp_nonce }}">
```

and also:

```
{% load csp %}
{% script %}
<script>
// ...
</script>
{% endscript %}
```

And none of them work.

I noted that the CSP header does contain the nonce.

I'm using Django 3.2 and Python 3.8.

**Comment by Flimm** [comment]
I'm not sure why, but all of a sudden, `<script nonce="{{ request.csp_nonce }}">` does render a nonce. However, it does not match the nonce in the CSP HTTP header.

I suspect these lines in CSP's middleware may be the culprit:

```
    def process_request(self, request):
        nonce = partial(self._make_nonce, request)
        request.csp_nonce = SimpleLazyObject(nonce)
```

Maybe Django's system doesn't play well with `SimpleLazyObject` as expected?

**Comment by chocolatkey** [comment]
I am having the same issue, python 3.9 and Django 3.2, the nonce is completely empty

**Comment by 9mido** [comment]
Can you try testing by rolling back to previous versions of django-csp, python, and django and see if the nonce still is not populating?

**Comment by tkthanikai** [comment]
Hi, I'm running into the same issue. I tested with previous versions of django-csp and it didn't help.

**Comment by tolexy** [comment]
I am seeing a similar issue: The context processor renders the nonce in the template and also appends the correct nonce value to the to the CSP Header, but script still blocked. Perhaps insertion/render order in template vs headers?

Header: 'nonce-Y3a0i+Nlg4oGsFOijreMvg=='   (appends to end of script-src)
Template: <script nonce="{{CSP_NONCE}}"     (same result with {{ request.csp_nonce }}")
Page source: <script nonce="Y3a0i+Nlg4oGsFOijreMvg=="
CSP_INCLUDE_NONCE_IN=['script-src','style-src',]

The script is still blocked. I tried moving the script out of <head> to just before </html>, using only one instance of {{ CSP_NONCE }} in html all to no avail. It still prints the correct nonce, but does not work. 

Moved csp middleware to top of list, did not help. 
Added csp in INSTALLED_APPs, did not help. 

Django 3.2.5, python 3.9.5, django-csp==3.7 : failed
Django 2.2.24 py 3.8.11, csp 3.7 : failed but this may be for other reasons in my env that would take too long to modify at this point. 

will try modifying the csp code in my env and see what happens...

**Comment by tolexy** [comment]
Following up on my previous post, it truly seems to be working, however Chrome and Firefox still block the scripts (specifically, this is for google analytics, but does the same on other scripts as well that try to use a nonce). From another comment on another issue here, it appears to be a browser issue, or perhaps a CSP spec issue. I havent been able to find a clear answer and unfortunately have to use unsafe-inline until this gets sorted.  It does now throw any errors/block scripts when running on local, only once deployed.
not sure if this cookie_flag is causing issues. it was required to get GA working in firefox...
 gtag('config', 'G-xxxxxx', { cookie_domain: 'xxxxxxxxx.com', cookie_flags: 'SameSite=None;Secure' });

**Comment by g-k** [comment]
Huh, weird. Closing based on https://github.com/mozilla/django-csp/issues/167#issuecomment-873099528 but we can revisit if it keeps happening.

**Comment by rlskoeser** [comment]
Sharing what I discovered with a similar problem, in case anyone trips over this and finds this issue.

When I inspected my script element using browser developer tools (Firefox or Chrome), it looked like there was a nonce attribute with no value. But when I used view source on the same page, the nonce value was present.

What I discovered is that the nonce is hidden from the DOM for security reasons — otherwise the security could be bypassed.  See discussion on https://bugzilla.mozilla.org/show_bug.cgi?id=1374612
